### ISSUE TITLE
Twisted port of the Python driver's connection class

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#3785]
This is a follow-up of https://github.com/rethinkdb/rethinkdb/issues/3298 .

> We will be providing an `on(cb)` function on cursors, and integrate with the framework's event loop to call the callback once we get a response.

This issue is for a Twisted-based implementation. Unless Twisted provides another natural way of handling errors, we should probably just make the cb a function on `(error, row)` like in the JS driver.

An open question is how to package the Twisted driver variant. The easiest way probably is to have a separate module 
```py
import rethinkdb-twisted as r
```
that is equivalent to `rethinkdb` except for using Twisted for i/o.

Another alternative is to have a special connection type, e.g. `r.connect(host, backend="twisted")`. That's a bit nicer to use I think, but I don't know how much work it is to get the driver module import properly on a system that doesn't even have Twisted installed.

Pinging @gchpaco @timmaxw @larkost for comments.

**Comment by danielmewes** [comment]
Also note https://github.com/rethinkdb/rethinkdb/issues/2622#issuecomment-74411502 regarding multiple backend.

If at all possible I think we should provide the `r.connect(host, backend="twisted")` interface (or something similar to that) rather than a separate module.

Calling `on` on a cursor that's on a connection using the default (synchronous) backend should just throw an error I think.

**Comment by danielmewes** [comment]
Alternatively we could support Tornado first (https://github.com/rethinkdb/rethinkdb/issues/2622#issuecomment-74560593). The same questions regarding the API and module interface as here would apply.

**Comment by danielmewes** [comment]
Decision is in: We're going to work on the Tornado integration first, and then see if we have time left to do Twisted for 2.0 as well or if we will wait for that until a little later.

**Comment by RaitoBezarius** [comment]
Are there any updates about this issue? I can see that now, it's planned for 2.0.x or 2.1?
I wanted to know what would be needed to make it work, except implementing a net_twisted.py (like net_tornado.py) ?

Thanks!

**Comment by danielmewes** [comment]
Hi @RaitoBezarius, when we are going to ship this will depend on how much demand there is for this backend.
Implementing it should just need a `net_twisted.py`, and a few tests for it (like the one in `test/rql_test/connections/tornado_connection.py`). We're happy to support you if you want to give it a shot!
For another data point, see this pull request by @takluyver for a Python 3 async backend: https://github.com/rethinkdb/rethinkdb/pull/4071 .

**Comment by RaitoBezarius** [comment]
Thanks @danielmewes! Is there some protocol spec or a general description of how a client driver should interact with the server and what kind of assumptions can the client make? (Anyway, I can just use those implementation details to work with).

**Comment by danielmewes** [comment]
@RaitoBezarius Very cool. I'm looking forward to your pull request. Feel free to put it up even before it's completely ready If you like, so we can give you early feedback.

We have this article on how the protocol works http://rethinkdb.com/docs/writing-drivers/ . I think for porting to Twisted you probably won't need most of this, since those things are taken care of in the higher levels of the driver (not in the `net_*.py` file).

**Comment by coffeemug** [comment]
IMO `r.connect(backend=X)` is better than a separate module, but what happens if people start two connections with different backends? Will they interop properly?

One other option is `r.set_async_backend(X)`, which I think is a good choice. Yet another option is `import rethinkdb.twisted as r`, which I think is more pythonesque than `import rethinkdb_twisted as r`.

**Comment by danielmewes** [comment]
Actually I didn't intend this to be a ReQL_proposal. We already have `set_async_backend` which we're going to use for all the new Python backends for the time being.
I'd prefer to not discuss the specifics of the Twisted API as part of the ReQL discussion period.

Sorry this slipped through. Removing the ReQL_proposal label.

**Comment by danielmewes** [comment]
`set_loop_type` I meant, not `set_async_backend` (but it's the same thing).
http://rethinkdb.com/api/python/set_loop_type/

**Comment by RaitoBezarius** [comment]
@danielmewes I wanted to ask you about the behavior of the cursors, I think that they are not very very async-friendly.

```python
    def _maybe_fetch_batch(self):
        if self.error is None and \
           len(self.items) <= self.threshold and \
           self.outstanding_requests == 0:
            self.outstanding_requests += 1
            self.conn._parent._continue(self)
```
Currently, _continue calls run_query which should return a Deferred/Future/Whatever.

But as we don't return it from `_maybe_fetch_batch`, thus, it is impossible to get the deferred object.
Furthermore, I think that there is a biggest problem (correct me if what I'll say is wrong).

Cursor works by waiting for the connection to extend them to get a new response.
In order to iterate over a cursor, we wait that those items are extended by the connection.
But how we wait during iteration: `while len(self.items) == 0`. With an async backend, it's maybe stupid to it like that, we should just wait for the deferred from _maybe_fetch_batch returns or sleep until we're extended:

```python
def _get_next(self, timeout):
    start_timeout(timeout, self.new_response)
    yield self.new_response
    returnValue(convert_pseudo(self.items.pop(0), self.query))
```

In fact, all _maybe_fetch_batch is to send CONTINUE query if I understand correctly?

It would clearly better if we implemented cursor just as a generator, by subclassing __iter__ and __next__, leading to a code (independent from any backend) : 
```python
for change in cursor:
    print change
```

It was a lot of thoughts in one message, sorry for it!

**Comment by Tryneus** [comment]
@RaitoBezarius, a few things:

You're right that `_maybe_fetch_batch` receives a future in an async environment, but at the moment it is always immediately available, so it doesn't cause any problems.  The CONTINUE query is sent with `noreply=True`, and it returns `None`.  We don't wait on these futures when sending the CONTINUE query because the user should only wait when consuming items from the cursor, not when creating it.

The `DefaultCursor` implements iteration, but we had problems getting iteration to work on `TornadoCursor` because Tornado coroutines are generators - and a `StopIteration` means that the coroutine is finished and has no more futures to process.  I couldn't find a safe way to have an iterator implemented in terms of a Tornado coroutine.

Looking forward to your PR, let me know if you have any questions!

**Comment by RaitoBezarius** [comment]
@Tryneus Thanks for your answer.
I've managed to get the iteration working here: https://github.com/RaitoBezarius/rethinkdb/commit/3abeec46e5e7a0dc4944fd4d91d2216768967fca
But, it yields deferred, and all I need to do is to yield them again.

As for tests, should I implement as many tests as there is for Tornado and Asyncio?
I'll open the PR.

**Comment by danielmewes** [comment]
@RaitoBezarius 's implementation has been merged and will ship with RethinkDB 2.1.
https://github.com/rethinkdb/rethinkdb/pull/4096
